### PR TITLE
Run publish on failed or succeeded

### PR DESCRIPTION
**Pull request by benbp**

### CODE REVIEW

File: eng/common/pipelines/templates/steps/publish-1es-artifact.yml [hunk 1]
```diff
@@ -23,7 +23,7 @@ steps:
     displayName: Set Failed Artifact Name
```
Comment (weshaggard) [inline eng/common/pipelines/templates/steps/publish-1es-artifact.yml:23]: See my other comment https://github.com/Azure/azure-sdk-tools/pull/7758#discussion_r1511884105 I think we can make this one step to set the artifact name.

Comment (benbp) [reply eng/common/pipelines/templates/steps/publish-1es-artifact.yml:23]: Ah excellent

File: eng/common/pipelines/templates/steps/publish-1es-artifact.yml [hunk 2]
```diff
@@ -23,7 +23,7 @@ steps:
...
 
   - task: 1ES.PublishPipelineArtifact@1
-    condition: and(succeeded(), ${{ parameters.CustomCondition }})
+    condition: and(failedOrSucceeded(), ${{ parameters.CustomCondition }})
```
Comment (scbedd) [inline eng/common/pipelines/templates/steps/publish-1es-artifact.yml:26]:
```suggestion
    condition: and(succeededOrFailed(), ${{ parameters.CustomCondition }})
```

[Ref](https://learn.microsoft.com/en-us/azure/devops/pipelines/process/expressions?view=azure-devops#succeededorfailed)

**Review by scbedd — APPROVED** [review]
Thanks!

File: eng/common/pipelines/templates/steps/publish-1es-artifact.yml [hunk 3]
```diff
@@ -5,26 +5,28 @@
...
-      Write-Host "##vso[task.setvariable variable=PublishArtifactName;]${{ parameters.ArtifactName }}-FailedAttempt$(System.JobAttempt)"
-    condition: and(failed(), ${{ parameters.CustomCondition }})
-    displayName: Set Failed Artifact Name
+      if ($env:AGENT_JOBSTATUS -eq "Failed") {
+        Write-Host "##vso[task.setvariable variable=PublishArtifactName;]${{ parameters.ArtifactName }}-FailedAttempt$(System.JobAttempt)"
+      } else {
+        Write-Host "##vso[task.setvariable variable=PublishArtifactName;]${{ parameters.ArtifactName }}"
+      }
+    condition: and(succeededOrFailed(), ${{ parameters.CustomCondition }})
+    displayName: Set Artifact Name $(Agent.JobStatus)
```
Comment (weshaggard) [inline eng/common/pipelines/templates/steps/publish-1es-artifact.yml:24]: I don't expect this variable to actually correctly render as it is runtime computed.

**Pull request merged**